### PR TITLE
Setting up models

### PR DESCRIPTION
**Pull request by lawsonbh**
Doing a first pass at setting up the relationships between our core models

Questions for @bbelderbos:
1. Does the Ego model need a many to one relationship with user from django? IE one user can create many egos 
2. A single ego can have many morphs - this means that one ego can have many character sheets IE for one adventure I may have my ego in splicer morph and in another adventure want to have my ego in a futura morph --- what's the best way to represent that? A many to many relationship between character_sheet and?

### CODE REVIEW

**Comment by lawsonbh** [comment]
I am also going to dig in tonight / tomorrow on understanding how to populate fields in other tables based on something in another model IE . . . the character sheet flex pool is . . . 
1. The ego's flex bonus
2. add the morph's flex bonus
3. add in additional flex bonus from customization points spent

File: bodybank/character_generator/models.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    ego_a_rep = models.IntegerField("@-rep")
+    ego_c_rep = models.IntegerField("c-rep")
+    ego_f_rep = models.IntegerField("f-rep")
+    ego_g_rep = models.IntegerField("g-rep")
+    ego_i_rep = models.IntegerField("i-rep")
+    ego_r_rep = models.IntegerField("r-rep")
+    ego_x_rep = models.IntegerField("x-rep")
+
+class Morph(models.Model):
+    egos = models.ManyToManyField(Ego)
```
Comment (bbelderbos) [inline bodybank/character_generator/models.py]: I wonder if both ego and morph should be linked in a new table called UserEgo, let's discuss in a bit ...

File: bodybank/character_generator/models.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    morph_name = models.CharField(max_length=50)
+    morph_type = models.CharField(max_length=50)
+    morph_cost = models.IntegerField()
+
+
+class Item(models.Model):
+    egos = models.ManyToManyField(Ego)
```
Comment (bbelderbos) [inline bodybank/character_generator/models.py]: Are items linked to an ego or a morph or are they independent? In latter case similar to UserEgo we might need to track them in that table rather. Let's discuss shortly ...

File: bodybank/character_generator/models.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
+from django.db import models
+
+class Ego(models.Model):
+    ego_name = models.CharField(max_length=50)
+    ego_cog_apt = models.IntegerField("Cognition (COG)")
+    ego_int_apt = models.IntegerField("Intuition (INT)")
+    ego_ref_apt = models.IntegerField("Reflex (REF)")
+    ego_sav_apt = models.IntegerField("Savy (SAV)")
+    ego_som_apt = models.IntegerField("Somatics (SOM)")
```
Comment (bbelderbos) [inline bodybank/character_generator/models.py]: "ego" is already part of the model name, what if we call the columns full words (cognition, intuition, reflex, etc), maybe more readable?

**Comment by bbelderbos** [comment]
> Doing a first pass at setting up the relationships between our core models
> 
> Questions for @bbelderbos:
> 
> 1. Does the Ego model need a many to one relationship with user from django? IE one user can create many egos
> 2. A single ego can have many morphs - this means that one ego can have many character sheets IE for one adventure I may have my ego in splicer morph and in another adventure want to have my ego in a futura morph --- what's the best way to represent that? A many to many relationship between character_sheet and?

1. Good question, would the end user be able to create many egos?
2. So then morph would be a manytomany on UserEgo? As per review and 1., I was thinking of this new table so that the logged in user can make multiple egos

**Comment by bbelderbos** [comment]
> 

When I got to the bonus part of the pdf I thought "can we defer this part"? It will get complex quickly. 

Also I think we need to keep the goal of the app in mind: it's not to simulate the game, right? People will play it outside of this app. This app is merely to track the data, no?

**Comment by lawsonbh** [comment]
Correct! This whole thing is just on generating characters - in fact that pdf I sent you gives the players pre-generated characters. Traditionally, this has been a big stumbling block for folks!

User logs in and can create a character or retrieve one that they have already created in the past.

**Comment by bbelderbos** [comment]
Ignore the notes on the PR, glad we had a good discussion and are on the same page. Morph + Item + User, they all come together on Ego: Morph + User as foreign keys (one to many) and Item as ManyToMany field. I think you took some good notes so looking forward to reviewing tomorrow. 

forms and views example is in the starter training:
https://github.com/pybites/django-starter/blob/solution/blog/forms.py
https://github.com/pybites/django-starter/blob/solution/blog/views.py

**Comment by lawsonbh** [comment]
Working through what I think is an error on how I am including urls under bodybank/urls.py referencing the bodybank/character_generator/urls.py. After submitting an ego, the http redirects to /ego which doesn't exist.

File: bodybank/character_generator/views.py [hunk 4]
```diff
@@ -1,4 +1,17 @@
...
+from django.shortcuts import render
+from .forms import EgoForm
 
 def index(request):
     return HttpResponse("Welcome to the Body Bank!")
+
+def get_ego(request):
+    if request.method == 'POST':
+        form = EgoForm(request.POST)
+        if form.is_valid():
```
Comment (bbelderbos) [inline bodybank/character_generator/views.py:11]: You still need to save it to the db with `form.save()`

File: bodybank/character_generator/views.py [hunk 5]
```diff
@@ -1,4 +1,17 @@
...
+            return HttpResponseRedirect('/thanks/')
```
Comment (bbelderbos) [inline bodybank/character_generator/views.py]: I would use the same view, so fall through to the `render(...` below

File: bodybank/character_generator/views.py [hunk 6]
```diff
@@ -1,4 +1,17 @@
 from django.http import HttpResponse
+from django.shortcuts import render
+from .forms import EgoForm
 
 def index(request):
     return HttpResponse("Welcome to the Body Bank!")
+
+def get_ego(request):
+    if request.method == 'POST':
```
Comment (bbelderbos) [inline bodybank/character_generator/views.py:9]: good

**Review by bbelderbos — APPROVED** [review]

File: bodybank/character_generator/urls.py [hunk 7]
```diff
@@ -3,5 +3,6 @@
 from . import views
 
 urlpatterns = [
-        path('', views.index, name='index')
+        path('', views.index, name='index'),
+        path('ego/',views.get_ego,name='build-ego')
```
Comment (bbelderbos) [inline bodybank/character_generator/urls.py]: maybe make the view function name consistent? `build_ego()`

File: bodybank/character_generator/templates/ego.html [hunk 8]
```diff
@@ -1,8 +1,4 @@
-<head>
-    <title>The Ego Page</title>
-    <h1>Ego Page!</h1>
-    <h3>Hi there, thanks for visiting the ego page!</h3>
-</head>
-<body>
-
-</body>
+<form action="/ego/" method="post">
```
Comment (bbelderbos) [inline bodybank/character_generator/templates/ego.html]: I would not hardcode urls, you can leave it empty to default to same page or use `{% url ... %}`

**Comment by bbelderbos** [comment]
> Working through what I think is an error on how I am including urls under bodybank/urls.py referencing the bodybank/character_generator/urls.py. After submitting an ego, the http redirects to /ego which doesn't exist.

OK good progress, let's debug in a bit ...

**Pull request merged**